sklearn/impute/_knn.py

```diff
@@ -0,0 +1,271 @@
+        if weight_matrix is not None:
+            weight_matrix[np.isnan(weight_matrix)] = 0.0
+
+        # Retrieve donor values and calculate kNN average
+        donors = fit_X_col[potential_donors_idx].take(donors_idx)
+        donors_mask = mask_fit_X_col[potential_donors_idx].take(donors_idx)
+        donors = np.ma.array(donors, mask=donors_mask)
+
+        return np.ma.average(donors, axis=1, weights=weight_matrix).data
+
+    def fit(self, X, y=None):
+        """Fit the imputer on X.
+
+        Parameters
+        ----------
+        X : array-like, shape=(n_samples, n_features)
+            Input data, where `n_samples` is the number of samples and
+            `n_features` is the number of features.
+
+        Returns
+        -------
+        self : object
+        """
+
+        # Check data integrity and calling arguments
+        if not is_scalar_nan(self.missing_values):
+            force_all_finite = True
+        else:
+            force_all_finite = "allow-nan"
+            if self.metric not in _NAN_METRICS and not callable(self.metric):
+                raise ValueError(
+                    "The selected metric does not support NaN values")
+        if self.n_neighbors <= 0:
+            raise ValueError(
+                "Expected n_neighbors > 0. Got {}".format(self.n_neighbors))
+
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+
+        _check_weights(self.weights)
+        self._fit_X = X
+
+        return self
+
+    def transform(self, X):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            The input data to complete.
+
+        Returns
+        -------
+        X : array-like, shape = (n_samples, n_output_features)
+            The imputed dataset. `n_output_features` is the number of features
+            that is not always missing during `fit`.
+        """
+
+        check_is_fitted(self, ["_fit_X"])
+        if not is_scalar_nan(self.missing_values):
+            force_all_finite = True
+        else:
+            force_all_finite = "allow-nan"
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+
+        if X.shape[1] != self._fit_X.shape[1]:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed")
+
+        mask = _get_mask(X, self.missing_values)
+        if not np.any(mask):
+            return X
+
+        row_missing_idx = np.flatnonzero(mask.any(axis=1))
+
+        # Pairwise distances between receivers and fitted samples
+        dist = pairwise_distances(X[row_missing_idx, :], self._fit_X,
+                                  metric=self.metric,
+                                  squared=False,
```

Comment:
This will break for an arbitrary callable metric, won't it?

Reply:
If you are referring to `force_all_finite`, this has been updated to `force_all_finite` which is defined in reference to `self.missing_value`.

If you are referring to `missing_value`, the `metric` docstring specifies that `missing_values` need to be a keyword.

Reply:
Thought he was referring to "squared"?

Reply:
Ah yes. PR updated to remove "squared".